internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala

```diff
@@ -636,17 +636,18 @@ private class MRelationsNameHashing(
       productClassName = productClassName
     )
 
-  override private[inc] def addUsedName(className: String, name: UsedName): Relations =
+  private[inc] def addUsedNames(data: Relation[String, UsedName]): Relations = {
     new MRelationsNameHashing(
       srcProd,
       libraryDep,
       libraryClassName,
       internalDependencies = internalDependencies,
       externalDependencies = externalDependencies,
       classes,
-      names = names + (className, name),
+      names = if (names.forwardMap.isEmpty) data else names ++ data,
```

Comment:
This optimization could be pushed down into `sbt-util`.

Reply:
... although that repo is now archived so I guess `Zinc` will end up with a copy of a subset of its data structues to avoid a circular dependency. @eed3si9n what's the plan there?